### PR TITLE
Fix/plugin checks

### PR DESCRIPTION
**Pull request by brianhogg**
## Description
Various security warning/error fixes.

## How has this been tested?
Manually

## Checklist:
- [x] This PR requires and contains at least one changelog file. 
- [x] My code has been tested.
- [x] My code passes all existing automated tests. 
- [x] My code follows the LifterLMS Coding & Documentation Standards.

### CODE REVIEW

File: includes/admin/reporting/tabs/class.llms.admin.reporting.tab.courses.php [hunk 1]
```diff
@@ -97,16 +69,14 @@ public function output() {
 		if ( isset( $_GET['course_id'] ) ) {
 
 			if ( ! current_user_can( 'edit_post', llms_filter_input( INPUT_GET, 'course_id', FILTER_SANITIZE_NUMBER_INT ) ) ) {
-				wp_die( __( 'You do not have permission to access this content.', 'lifterlms' ) );
+				wp_die( esc_attr__( 'You do not have permission to access this content.', 'lifterlms' ) );
```
Comment (ideadude) [inline includes/admin/reporting/tabs/class.llms.admin.reporting.tab.courses.php]:
```suggestion
				wp_die( esc_html__( 'You do not have permission to access this content.', 'lifterlms' ) );
```

File: includes/admin/reporting/tabs/class.llms.admin.reporting.tab.memberships.php [hunk 2]
```diff
@@ -72,7 +70,7 @@ public function output() {
 		if ( isset( $_GET['membership_id'] ) ) {
 
 			if ( ! current_user_can( 'edit_post', llms_filter_input( INPUT_GET, 'membership_id', FILTER_SANITIZE_NUMBER_INT ) ) ) {
-				wp_die( __( 'You do not have permission to access this content.', 'lifterlms' ) );
+				wp_die( esc_attr__( 'You do not have permission to access this content.', 'lifterlms' ) );
```
Comment (ideadude) [inline includes/admin/reporting/tabs/class.llms.admin.reporting.tab.memberships.php]:
```suggestion
				wp_die( esc_html__( 'You do not have permission to access this content.', 'lifterlms' ) );
```

File: includes/admin/reporting/tabs/class.llms.admin.reporting.tab.students.php [hunk 3]
```diff
@@ -105,11 +103,11 @@ public function output() {
 
 			$student_id = llms_filter_input( INPUT_GET, 'student_id', FILTER_SANITIZE_NUMBER_INT );
 			if ( ! llms_current_user_can( 'view_lifterlms_reports', $student_id ) ) {
-				wp_die( __( "You do not have permission to access this student's reports", 'lifterlms' ) );
+				wp_die( esc_attr__( "You do not have permission to access this student's reports", 'lifterlms' ) );
```
Comment (ideadude) [inline includes/admin/reporting/tabs/class.llms.admin.reporting.tab.students.php]:
```suggestion
				wp_die( esc_html__( "You do not have permission to access this student's reports", 'lifterlms' ) );
```

File: includes/admin/reporting/tabs/class.llms.admin.reporting.tab.students.php [hunk 4]
```diff
@@ -105,11 +103,11 @@ public function output() {
...
 			}
 			$student = llms_get_student( $student_id );
 			if ( ! $student ) {
-				wp_die( __( "This student doesn't exist.", 'lifterlms' ) );
+				wp_die( esc_attr__( "This student doesn't exist.", 'lifterlms' ) );
```
Comment (ideadude) [inline includes/admin/reporting/tabs/class.llms.admin.reporting.tab.students.php]:
```suggestion
				wp_die( esc_html__( "This student doesn't exist.", 'lifterlms' ) );
```

File: includes/admin/settings/class.llms.settings.checkout.php [hunk 5]
```diff
@@ -221,7 +217,7 @@ private function get_settings_default() {
 			array(
 				'default' => 'yes',
 				'desc'    => __( 'Enable automatic retry of failed recurring payments.', 'lifterlms' ) .
-							  '<p class="description">' . sprintf( __( 'Recover lost revenue from temporarily declined payment methods. %1$sLearn More%2$s', 'lifterlms' ), '<a href="https://lifterlms.com/docs/automatic-retry-failed-payments/" target="_blank">', '</a>' ) . '</p>',
+								'<p class="description">' . sprintf( __( 'Recover lost revenue from temporarily declined payment methods. %1$sLearn More%2$s', 'lifterlms' ), '<a href="https://lifterlms.com/docs/automatic-retry-failed-payments/" target="_blank">', '</a>' ) . '</p>',
```
Comment (ideadude) [inline includes/admin/settings/class.llms.settings.checkout.php]:
```suggestion
								'<p class="description">' . sprintf( esc_html__( 'Recover lost revenue from temporarily declined payment methods. %1$sLearn More%2$s', 'lifterlms' ), '<a href="https://lifterlms.com/docs/automatic-retry-failed-payments/" target="_blank">', '</a>' ) . '</p>',
```

File: includes/admin/settings/class.llms.settings.checkout.php [hunk 6]
```diff
@@ -221,7 +217,7 @@ private function get_settings_default() {
...
 				'id'      => 'lifterlms_recurring_payment_retry',
 				'title'   => __( 'Retry Failed Payments', 'lifterlms' ),
```
Comment (ideadude) [inline includes/admin/settings/class.llms.settings.checkout.php:222]:
```suggestion
				'title'   => esc_html__( 'Retry Failed Payments', 'lifterlms' ),
```

Comment (ideadude) [reply includes/admin/settings/class.llms.settings.checkout.php:222]: Not sure if the plugin checker is finding these here or elsewhere where a var is set to a translated string, but we basically can use esc_html__ to sanitize the value coming from the translators.

File: includes/class-llms-generator-courses.php [hunk 7]
```diff
@@ -153,9 +151,9 @@ public function generate_course( $raw ) {
 	public function generate_courses( $raw ) {
 
 		if ( empty( $raw['courses'] ) ) {
-			throw new Exception( __( 'Raw data is missing the required "courses" array.', 'lifterlms' ), self::ERROR_GEN_MISSING_REQUIRED );
+			throw new Exception( esc_attr__( 'Raw data is missing the required "courses" array.', 'lifterlms' ), intval( self::ERROR_GEN_MISSING_REQUIRED ) );
```
Comment (ideadude) [inline includes/class-llms-generator-courses.php:154]:
```suggestion
			throw new Exception( esc_html__( 'Raw data is missing the required "courses" array.', 'lifterlms' ), intval( self::ERROR_GEN_MISSING_REQUIRED ) );
```

File: includes/class-llms-generator-courses.php [hunk 8]
```diff
@@ -153,9 +151,9 @@ public function generate_course( $raw ) {
...
 		} elseif ( ! is_array( $raw['courses'] ) ) {
-			throw new Exception( __( 'The raw "courses" item must be an array.', 'lifterlms' ), self::ERROR_GEN_INVALID_FORMAT );
+			throw new Exception( esc_attr__( 'The raw "courses" item must be an array.', 'lifterlms' ), intval( self::ERROR_GEN_INVALID_FORMAT ) );
```
Comment (ideadude) [inline includes/class-llms-generator-courses.php:156]:
```suggestion
			throw new Exception( esc_html__( 'The raw "courses" item must be an array.', 'lifterlms' ), intval( self::ERROR_GEN_INVALID_FORMAT ) );
```

File: includes/class-llms-generator-courses.php [hunk 9]
```diff
@@ -448,7 +441,7 @@ protected function create_question( $raw, $manager, $author_id ) {
 		);
 
 		if ( ! $question_id ) {
-			throw new Exception( __( 'Error creating the question post object.', 'lifterlms' ), self::ERROR_CREATE_POST );
+			throw new Exception( esc_attr__( 'Error creating the question post object.', 'lifterlms' ), intval( self::ERROR_CREATE_POST ) );
```
Comment (ideadude) [inline includes/class-llms-generator-courses.php:444]:
```suggestion
			throw new Exception( esc_html__( 'Error creating the question post object.', 'lifterlms' ), intval( self::ERROR_CREATE_POST ) );
```

File: includes/class.llms.achievement.php [hunk 10]
```diff
@@ -118,7 +118,7 @@ class LLMS_Achievement {
 	 * @return void
 	 */
 	public function __call( $name, $args ) {
-		_deprecated_function( __CLASS__ . '::' . $name, '6.0.0' );
+		_deprecated_function( __CLASS__ . '::' . esc_attr( $name ), '6.0.0' );
```
Comment (ideadude) [inline includes/class.llms.achievement.php]:
```suggestion
		_deprecated_function( __CLASS__ . '::' . esc_html( $name ), '6.0.0' );
```

File: includes/achievements/class.llms.achievement.user.php [hunk 11]
```diff
@@ -106,7 +106,7 @@ class LLMS_Achievement_User extends LLMS_Achievement {
 	 * @return void
 	 */
 	public function __call( $name, $args ) {
-		_deprecated_function( __CLASS__ . '::' . $name, '6.0.0' );
+		_deprecated_function( __CLASS__ . '::' . esc_attr( $name ), '6.0.0' );
```
Comment (ideadude) [inline includes/achievements/class.llms.achievement.user.php]:
```suggestion
		_deprecated_function( __CLASS__ . '::' . esc_html( $name ), '6.0.0' );
```

File: includes/class.llms.emails.php [hunk 12]
```diff
@@ -138,13 +138,12 @@ public function get_css( $rule = '', $echo = true ) {
 		if ( isset( $css[ $rule ] ) ) {
 
 			if ( $echo ) {
-				echo $css[ $rule ];
+				echo esc_attr( $css[ $rule ] );
```
Comment (ideadude) [inline includes/class.llms.emails.php:141]: I don't know what CSS rules are being used here, but I feel esc_attr could break certain styles yeah?

If we ensure earlier that the values used to generated the rules are sanitized, we can ignore the escape check here.

File: includes/class.llms.emails.php [hunk 13]
```diff
@@ -327,18 +324,16 @@ public function get_table_html( $rows ) {
 
 		ob_start();
 		?>
-		<table style="<?php echo $table_style; ?>">
+		<table style="<?php echo esc_attr( $table_style ); ?>">
```
Comment (ideadude) [inline includes/class.llms.emails.php:327]: How can we test this. I feel these styles and content may include things broken by the escaping functions.

File: includes/class.llms.install.php [hunk 14]
```diff
@@ -697,19 +687,18 @@ public static function update_actions() {
 		}
 
 		if ( ! llms_verify_nonce( 'llms-db-update', 'do_db_updates', 'GET' ) ) {
-			wp_die( __( 'Action failed. Please refresh the page and retry.', 'lifterlms' ) );
+			wp_die( esc_attr__( 'Action failed. Please refresh the page and retry.', 'lifterlms' ) );
```
Comment (ideadude) [inline includes/class.llms.install.php]:
```suggestion
			wp_die( esc_html__( 'Action failed. Please refresh the page and retry.', 'lifterlms' ) );
```

File: includes/class.llms.install.php [hunk 15]
```diff
@@ -697,19 +687,18 @@ public static function update_actions() {
...
 		}
 
 		if ( ! current_user_can( 'manage_options' ) ) {
-			wp_die( __( 'You are not allowed to perform the requested action.', 'lifterlms' ) );
+			wp_die( esc_attr__( 'You are not allowed to perform the requested action.', 'lifterlms' ) );
```
Comment (ideadude) [inline includes/class.llms.install.php]:
```suggestion
			wp_die( esc_html__( 'You are not allowed to perform the requested action.', 'lifterlms' ) );
```

File: includes/class.llms.post.relationships.php [hunk 16]
```diff
@@ -203,12 +201,11 @@ public static function maybe_prevent_product_deletion( $delete, $post ) {
 			add_filter( 'rest_request_after_callbacks', array( __CLASS__, 'rest_filter_products_with_active_subscriptions_error_message' ), 10, 3 );
 		} else { // Deleting via wp-admin.
 			wp_die(
-				self::delete_product_with_active_subscriptions_error_message( $product->get( 'id' ) )
+				esc_attr( self::delete_product_with_active_subscriptions_error_message( $product->get( 'id' ) ) )
```
Comment (ideadude) [inline includes/class.llms.post.relationships.php]:
```suggestion
				esc_html( self::delete_product_with_active_subscriptions_error_message( $product->get( 'id' ) ) )
```

File: includes/forms/class-llms-forms-unsupported-versions.php [hunk 17]
```diff
@@ -80,9 +78,9 @@ public function init() {
 	public function output_notice() {
 		?>
 		<div class="notice notice-error">
-			<p><b><?php _e( 'Minimum Version Requirements Error', 'lifterlms' ); ?></b></p>
-			<p><?php printf( __( 'In order to manage LifterLMS Forms you must upgrade to at least WordPress version %s or later or install the latest version of the Gutenberg plugin.', 'lifterlms' ), LLMS_Forms::instance()::MIN_WP_VERSION ); ?></p>
-			<p><?php _e( 'If you do not upgrade, your forms will display properly on the frontend and users will be able to create accounts, enroll, and checkout but you will be unable to customize them.', 'lifterlms' ); ?></p>
+			<p><b><?php esc_html_e( 'Minimum Version Requirements Error', 'lifterlms' ); ?></b></p>
+			<p><?php printf( esc_html__( 'In order to manage LifterLMS Forms you must upgrade to at least WordPress version %s or later or install the latest version of the Gutenberg plugin.', 'lifterlms' ), esc_attr( LLMS_Forms::instance()::MIN_WP_VERSION ) ); ?></p>
```
Comment (ideadude) [inline includes/forms/class-llms-forms-unsupported-versions.php]:
```suggestion
			<p><?php printf( esc_html__( 'In order to manage LifterLMS Forms you must upgrade to at least WordPress version %s or later or install the latest version of the Gutenberg plugin.', 'lifterlms' ), esc_html( LLMS_Forms::instance()::MIN_WP_VERSION ) ); ?></p>
```

**Review by ideadude — COMMENTED** [review]
Mostly minor stuff and some cases where I'd like to double check that we're sure there isn't supposed to be HTML in things we're using esc_html on. A few cases of CSS/style code that I believe breaks when escaped sometimes.

Some places where you used esc_attr__, where I would have used esc_html__. These functions are literally the same, except for the filter applied. But good to only use the attribute one for html attributes as far as I know.

https://github.com/WordPress/WordPress/blob/9702c2e265bd3aff0522f55063d97ca21cc70099/wp-includes/formatting.php#L4673-L4721

File: includes/abstracts/abstract.llms.admin.table.php [hunk 18]
```diff
@@ -538,66 +535,92 @@ protected function get_table_classes() {
...
+					<?php $this->output_table_filters_html(); ?>
 				<?php endif; ?>
 			</header>
 			<table
-				class="<?php echo implode( ' ', $classes ); ?>"
-				data-args='<?php echo wp_json_encode( $this->get_args() ); ?>'
-				data-handler="<?php echo $this->get_handler(); ?>"
-				id="llms-gb-table-<?php echo $this->id; ?>"
+				class="<?php echo esc_attr( implode( ' ', $classes ) ); ?>"
+				data-args='<?php echo esc_attr( wp_json_encode( $this->get_args() ) ); ?>'
```
Comment (brianhogg) [inline includes/abstracts/abstract.llms.admin.table.php:614]: We should be good to esc_attr this even though wp_json_encode() should be safe

File: templates/checkout/form-confirm-payment.php [hunk 19]
```diff
@@ -89,17 +89,17 @@
...
+				<h4 class="llms-form-heading"><?php esc_html_e( 'Payment Details', 'lifterlms' ); ?></h4>
 				<div class="llms-checkout-section-content llms-form-fields">
 
 					<div class="llms-payment-method">
 						<?php do_action( 'lifterlms_checkout_confirm_before_payment_method', $gateway_id ); ?>
-						<span class="llms-gateway-title"><span class="llms-label"><?php _e( 'Payment Method:', 'lifterlms' ); ?></span> <?php echo $selected_gateway->get_title(); ?></span>
+						<span class="llms-gateway-title"><span class="llms-label"><?php esc_html_e( 'Payment Method:', 'lifterlms' ); ?></span> <?php echo esc_html( $selected_gateway->get_title() ); ?></span>
 						<?php if ( $selected_gateway->get_icon() ) : ?>
-							<span class="llms-gateway-icon"><?php echo $selected_gateway->get_icon(); ?></span>
+							<span class="llms-gateway-icon"><?php echo wp_kses_post( $selected_gateway->get_icon() ); ?></span>
```
Comment (brianhogg) [inline templates/checkout/form-confirm-payment.php:99]: These changes or something up the chain are breaking some of the checkout functionality. The icons beside gateways don't appear (ie. credit card images, "powered by stripe" for credit card, and the first gateway isn't checked by default (but is bolded)

Comment (brianhogg) [reply templates/checkout/form-confirm-payment.php:99]: Adding additional tags and attributes to the LLMS_Form_Field::render() array for wp_kses() seems to have solved the output. It's part of the `icon` property on the form field which gets combined into the description, and it was being stripped out (`img` and an `a` tag)

File: includes/abstracts/abstract.llms.admin.table.php (hunk 18 above)
Comment (ideadude) [reply includes/abstracts/abstract.llms.admin.table.php:614]: Agreed. esc_attr and esc_html don't seem to interfere with the json as used here and similar places.

**Pull request merged**